### PR TITLE
Preprocessing

### PR DESCRIPTION
**Pull request by dwu359**
Allows the user to input preprocessing code to preprocess CSV data as a data frame through an AWS lambda function

### CODE REVIEW

File: frontend/playground-frontend/src/components/Home/CSVInputFile.js [hunk 1]
```diff
@@ -1,15 +1,16 @@
-import React, { useState } from "react";
+import React from "react";
 import PropTypes from "prop-types";
 import * as XLSX from "xlsx";
 import { FaCloudUploadAlt } from "react-icons/fa";
+//import { getSignedUploadUrl } from "../helper_functions/TalkWithBackend";
```
Comment (karkir0003) [inline frontend/playground-frontend/src/components/Home/CSVInputFile.js:5]: commented out code. is this needed? @dwu359

File: frontend/playground-frontend/src/components/Home/CSVInputFile.js [hunk 2]
```diff
@@ -1,15 +1,16 @@
...
+//import axios from "axios";
```
Comment (karkir0003) [inline frontend/playground-frontend/src/components/Home/CSVInputFile.js:6]: dead code? @dwu359

File: frontend/playground-frontend/src/components/Home/CSVInputFile.js [hunk 3]
```diff
@@ -45,28 +46,67 @@ const CSVInputFile = (props) => {
...
+    console.log(response);
+    const url = new URL(response["url"] + response["fields"]["key"]);
+    url.searchParams.append(
+      "AWSAccessKeyId",
+      response["fields"]["AWSAccessKeyId"]
+    );
+    url.searchParams.append("policy", response["fields"]["policy"]);
+    url.searchParams.append("signature", response["fields"]["signature"]);
+
+    console.log(
```
Comment (karkir0003) [inline frontend/playground-frontend/src/components/Home/CSVInputFile.js:97]: dead code? is this needed? @dwu359

File: tests/test_preprocessing.py [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+        """import pandas as pd\ndef preprocess(df):\n\tdf = df.drop(columns=['variety'])\n\treturn df""",
+    ],
+)
+def test_drop_column(read_csv_to_json, code_snippet):
+    payload = json.dumps({
+            "data": read_csv_to_json,
+            "code": code_snippet
+    })
+    resJson = invoke('preprocess_data', payload)
+    assert "columns" in resJson and resJson['columns'] == ['sepal.length', 'sepal.width', 'petal.length', 'petal.width'], resJson
```
Comment (karkir0003) [inline tests/test_preprocessing.py:37]: even simpler you can write `assert "columns" in resJson and "variety" not in resJson['columns']` @dwu359

File: tests/test_preprocessing.py [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+
+@pytest.mark.parametrize(
+    "code_snippet",
+    [
+        """import pandas as pd\ndef preprocess(df):\n\tdf['test'] = df['sepal.length'] + df['sepal.width']\n\treturn df""",
+    ],
+)
+def test_add_two_columns(read_csv_to_json, code_snippet):
+    payload = json.dumps({
```
Comment (karkir0003) [inline tests/test_preprocessing.py:46]: you can probably write a test case to see if Label Encoding on the "variety" column works properly. see https://www.kaggle.com/code/rajmehra03/the-iris-species-dataset @dwu359

File: tests/test_preprocessing.py [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+            "data": read_csv_to_json,
+            "code": code_snippet
+    })
+    resJson = invoke('preprocess_data', payload)
+    assert "columns" in resJson and resJson['columns'] == ['sepal.length', 'sepal.width', 'petal.length', 'petal.width', 'variety', 'test'], resJson
```
Comment (karkir0003) [inline tests/test_preprocessing.py:51]: test case to drop rows containing nan (ie: `dropna()`) function in pandas. see if the number of rows matches up with expected (ie: try doing dropna in google colab on iris dataset and see what the number of rows is) @dwu359

**Review by karkir0003 — CHANGES_REQUESTED** [review]
minor cleanups and comments

File: frontend/playground-frontend/src/components/Home/CSVInputFile.js [hunk 7]
```diff
@@ -92,6 +132,9 @@ const CSVInputFile = (props) => {
 CSVInputFile.propTypes = {
   setData: PropTypes.func.isRequired,
   setColumns: PropTypes.func.isRequired,
+  setOldData: PropTypes.func.isRequired,
+  fileName: PropTypes.any,
```
Comment (farisdurrani) [inline frontend/playground-frontend/src/components/Home/CSVInputFile.js:136]: Please declare the type of fileName. Maybe `PropTypes.string`?

Comment (dwu359) [reply frontend/playground-frontend/src/components/Home/CSVInputFile.js:136]: @farisdurrani I tried doing that but I don't think PropTypes.string allows null values, causing a compile error, so I had to use PropTypes.any to resolve that

Comment (farisdurrani) [reply frontend/playground-frontend/src/components/Home/CSVInputFile.js:136]: That's odd. Not putting `.isRequired` should allow null values but I trust you

**Review by farisdurrani — APPROVED** [review]

**Review by karkir0003 — APPROVED** [review]

**Pull request merged**